src/writer.js

```diff
@@ -182,6 +167,82 @@ class Writer extends EventEmitter {
   close() {
     return this.conn.destroy();
   }
+
+  _serializeMsg(msg){
+    if (_.isString(msg) || Buffer.isBuffer(msg)) {
+      return msg;
+    }
+    return JSON.stringify(msg);
+  }
+
+  _checkStateValidity(){
+    let connState = '';
+
+    if (this.conn && this.conn.statemachine) {
+      connState = this.conn.statemachine.current_state_name;
+    }
+
+    if (!this.conn || ['CLOSED', 'ERROR'].includes(connState)) {
+      return new Error('No active Writer connection to send messages');
+    }
+  }
+
+
+
+  _checkMsgsValidity(msgs){
+    // maybe when an array check every message to not be empty
+    if (!msgs || _.isEmpty(msgs)) {
+      return new Error('Attempting to publish an empty message');
+    }
+  }
+
+
+  _checkTimeMsValidity(timeMs){
+    return _.isNumber(timeMs)? undefined : new Error('The Delay must be a number') ;
+  }
+
+  _throwOrCallback(err, callback){
+    if (callback) {
+        return callback(err);
+      }
+      throw err;  
+  }
+
+
+  _callwhenReady(){
+    let args = arguments;
+    let isDeferPublish = arguments.length === 4;
```

Comment:
Instead of using the args length to determine the function to call, how about explicitly passing in the function to call on ready?